File: lab/2/main.py
import numpy as np
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


class Perceptron:
    def __init__(self, input_size=30, lr=0.01) -> None:
        self.input_size = input_size
        self.threshold = np.random.rand()
        self.w = np.random.rand(self.input_size, 1)
        self.lr = lr

    def forward(self, x):
        # Linear combination
        linear_output = np.dot(x, self.w) + self.threshold
        return self.sigmoid(linear_output)

    def sigmoid(self, out):
        return 1 / (1 + np.exp(-out))

    def sigmoid_derivative(self, out):
        # Derivative of the sigmoid function
        return out * (1 - out)

    def backward(self, x, y_true, y_pred):
        # Compute the error
        error = y_pred - y_true.reshape(-1, 1)

        # Derivative of sigmoid (for the output layer)
        d_pred = error * self.sigmoid_derivative(y_pred)

        # Update weights and threshold (bias)
        self.w -= self.lr * np.dot(x.T, d_pred)  # Gradient w.r.t weights
        # Gradient w.r.t threshold (bias)
        self.threshold -= self.lr * np.sum(d_pred)

    def train(self, x, y_true, epochs=100):
        for epoch in range(epochs):
            y_pred = self.forward(x)
            self.backward(x, y_true, y_pred)

    def predict(self, x):
        y_pred = self.forward(x)
        # Threshold at 0.5 to get binary output
        return (y_pred > 0.5).astype(int)


# Load the dataset
cancer_data = load_breast_cancer()

# Features (input data)
X = cancer_data.data

# Labels (output data)
y = cancer_data.target

# Normalize the data for faster convergence
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42)

# Initialize the Perceptron
perceptron = Perceptron(input_size=X_train.shape[1], lr=0.01)

# Train the Perceptron
perceptron.train(X_train, y_train, epochs=1000)

# Predict on the test set
y_pred = perceptron.predict(X_test)

# Calculate accuracy
accuracy = np.mean(y_pred.flatten() == y_test)
print(f"Accuracy: {accuracy * 100:.2f}%")

File: lab/2/main.m
% Load the dataset
data = readtable('breast_cancer_data.csv'); % Load your dataset here
X = data{:, 1:end-1}; % Features
y = data{:, end}; % Labels

% Normalize the data
X = zscore(X);

% Split the dataset into training and test sets
cv = cvpartition(size(X, 1), 'HoldOut', 0.2);
X_train = X(training(cv), :);
y_train = y(training(cv), :);
X_test = X(test(cv), :);
y_test = y(test(cv), :);

% Hyperparameters
input_size = size(X_train, 2);
lr = 0.001;
epochs = 50;

% Initialize weights and threshold
w = rand(input_size, 1);
threshold = rand();

% Train the Perceptron
for epoch = 1:epochs
    y_pred_train = forward(X_train, w, threshold);
    train_loss = compute_loss(y_train, y_pred_train);
    
    % Update weights and threshold
    [w, threshold] = backward(X_train, y_train, y_pred_train, w, threshold, lr);
end

% Predict on the test set
y_pred1 = predict(X_test, w, threshold);
accuracy = mean(y_pred1 == y_test);
fprintf('Accuracy (lr:0.001): %.2f%%\n', accuracy * 100);

% Function to compute forward pass
function out = forward(x, w, threshold)
    linear_output = x * w + threshold;
    out = sigmoid(linear_output);
end

% Function to compute sigmoid
function out = sigmoid(x)
    out = 1 ./ (1 + exp(-x));
end

% Function to compute loss
function loss = compute_loss(y_true, y_pred)
    loss = 0.5 * mean((y_pred - y_true).^2);
end

% Function to perform backward pass and update weights
function [w, threshold] = backward(x, y_true, y_pred, w, threshold, lr)
    error = y_pred - y_true;
    d_pred = error .* (y_pred .* (1 - y_pred));
    w = w - lr * (x' * d_pred);
    threshold = threshold - lr * sum(d_pred);
end

% Function to make predictions
function y_pred = predict(x, w, threshold)
    y_pred = forward(x, w, threshold) > 0.5;
end
